File: storm-metamodel-processor/src/main/java/st/orm/metamodel/TypeIndexProcessor.java
/*
 * Copyright 2024 - 2026 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package st.orm.metamodel;

import javax.annotation.processing.*;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.*;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import javax.tools.FileObject;
import javax.tools.StandardLocation;
import java.io.IOException;
import java.io.Writer;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.Set;

import static javax.tools.Diagnostic.Kind.ERROR;
import static javax.tools.Diagnostic.Kind.NOTE;

@SupportedSourceVersion(SourceVersion.RELEASE_21)
@SupportedAnnotationTypes("*")
public class TypeIndexProcessor extends AbstractProcessor {

    private Types types;
    private Elements elements;
    private Filer filer;

    private static final Set<String> INDEXED_TYPES = Set.of(
            "st.orm.Data",
            "st.orm.Converter"
    );

    private static final String INDEX_DIR = "META-INF/storm/";
    private final Map<String, TypeMirror> indexedTypeMirrors = new LinkedHashMap<>();
    private final Map<String, Set<String>> indexEntries = new LinkedHashMap<>();

    @Override
    public synchronized void init(ProcessingEnvironment processingEnv) {
        super.init(processingEnv);
        this.types = processingEnv.getTypeUtils();
        this.elements = processingEnv.getElementUtils();
        this.filer = processingEnv.getFiler();
        // Resolve all configured types to TypeMirrors (if present on the classpath).
        for (String typeFqName : INDEXED_TYPES) {
            TypeElement typeElement = elements.getTypeElement(typeFqName);
            if (typeElement != null) {
                indexedTypeMirrors.put(typeFqName, typeElement.asType());
            }
            // Always create an entry set, even if the type isn't present.
            indexEntries.put(typeFqName, new LinkedHashSet<>());
        }
    }

    @Override
    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
        processingEnv.getMessager().printMessage(NOTE, "Storm Type Index Processor is running.");
        if (indexedTypeMirrors.isEmpty()) {
            return false;
        }
        for (Element element : roundEnv.getRootElements()) {
            if (!(element instanceof TypeElement typeElement)) {
                continue;
            }
            String fqcn = elements.getBinaryName(typeElement).toString();
            for (Map.Entry<String, TypeMirror> entry : indexedTypeMirrors.entrySet()) {
                String typeFqName = entry.getKey();
                TypeMirror targetType = entry.getValue();
                if (isSubtypeOf(typeElement, targetType)) {
                    indexEntries.get(typeFqName).add(fqcn);
                }
            }
        }
        // After all rounds, write the index files.
        if (roundEnv.processingOver()) {
            for (Map.Entry<String, Set<String>> entry : indexEntries.entrySet()) {
                String typeFqName = entry.getKey();
                Set<String> types = entry.getValue();
                writeIndex(typeFqName, types);
            }
        }
        return false; // Do not claim any annotations.
    }

    private boolean isSubtypeOf(TypeElement type, TypeMirror target) {
        if (type.getKind() != ElementKind.CLASS) {
            return false;
        }
        TypeMirror a = types.erasure(type.asType());
        TypeMirror b = types.erasure(target);
        if (types.isSameType(a, b)) {
            return false;
        }
        return types.isSubtype(a, b);
    }

    private void writeIndex(String typeFqName, Set<String> types) {
        if (types.isEmpty()) {
            return;
        }
        String resourceName = INDEX_DIR + typeFqName + ".idx";
        try {
            FileObject file = filer.createResource(
                    StandardLocation.CLASS_OUTPUT,
                    "",
                    resourceName
            );
            try (Writer writer = file.openWriter()) {
                for (String fqcn : types) {
                    writer.write(fqcn);
                    writer.write('\n');
                }
            }
        } catch (IOException e) {
            processingEnv.getMessager().printMessage(
                    ERROR,
                    "Failed to write index for " + typeFqName + ". Error: " + e
            );
        }
    }
}

File: storm-metamodel-ksp/src/main/kotlin/st/orm/metamodel/TypeIndexProcessor.kt
/*
 * Copyright 2024 - 2026 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package st.orm.metamodel

import com.google.devtools.ksp.getAllSuperTypes
import com.google.devtools.ksp.processing.CodeGenerator
import com.google.devtools.ksp.processing.Dependencies
import com.google.devtools.ksp.processing.KSPLogger
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.processing.SymbolProcessor
import com.google.devtools.ksp.symbol.ClassKind
import com.google.devtools.ksp.symbol.KSAnnotated
import com.google.devtools.ksp.symbol.KSClassDeclaration

class TypeIndexProcessor(
    private val codeGenerator: CodeGenerator,
    private val logger: KSPLogger,
) : SymbolProcessor {

    private val indexEntries: MutableMap<String, MutableSet<String>> =
        INDEXED_TYPES.associateWith { linkedSetOf<String>() }.toMutableMap()

    override fun process(resolver: Resolver): List<KSAnnotated> {
        logger.info("Storm Type Index KSP is running.")
        resolver.getAllFiles().forEach { file ->
            file.declarations
                .filterIsInstance<KSClassDeclaration>()
                .forEach { clazz ->
                    val fqName = clazz.qualifiedName?.asString() ?: return@forEach
                    INDEXED_TYPES.forEach { typeFqName ->
                        if (clazz.isSubtypeOfFqName(typeFqName)) {
                            indexEntries.getValue(typeFqName) += fqName
                        }
                    }
                }
        }
        return emptyList()
    }

    override fun finish() {
        indexEntries.forEach { (typeFqName, lines) ->
            writeIndex(typeFqName, lines)
        }
    }

    private fun writeIndex(typeFqName: String, lines: Set<String>) {
        if (lines.isEmpty()) return
        val file = codeGenerator.createNewFile(
            Dependencies.ALL_FILES,
            INDEX_PACKAGE,
            typeFqName,
            "idx",
        )
        file.bufferedWriter().use { out ->
            lines.sorted().forEach { fqcn ->
                out.appendLine(fqcn)
            }
        }
    }

    companion object {
        private const val INDEX_PACKAGE = "META-INF.storm"

        private val INDEXED_TYPES = listOf(
            "st.orm.Data",
            "st.orm.Converter",
        )
    }
}

private fun KSClassDeclaration.isSubtypeOfFqName(targetFqName: String): Boolean {
    if (classKind != ClassKind.CLASS) return false
    val selfFqName = qualifiedName?.asString() ?: return false
    if (selfFqName == targetFqName) return false
    return getAllSuperTypes().any { superType ->
        superType.declaration.qualifiedName?.asString() == targetFqName
    }
}
